lib/BasicAnalyses/CSVAliasAnalysis.cpp

```diff
@@ -0,0 +1,259 @@
+    std::vector<Metadata *> OtherCSVScopes;
+    for (const auto &Q : CSVAliasInfoMap)
+      if (Q.first != CSV)
+        OtherCSVScopes.push_back(Q.second.AliasScope);
+
+    MDNode *CSVNoAliasSet = MDNode::get(Context, OtherCSVScopes);
+    AliasInfo.NoAliasSet = CSVNoAliasSet;
+  }
+}
+
+template<bool SegregateStackAccesses>
+void CSVAAP<SegregateStackAccesses>::decorateMemoryAccesses(Instruction &I) {
+  Value *Ptr = nullptr;
+
+  if (auto *L = dyn_cast<LoadInst>(&I))
+    Ptr = L->getPointerOperand();
+  else if (auto *S = dyn_cast<StoreInst>(&I))
+    Ptr = S->getPointerOperand();
+  else
+    return;
+
+  // Check if the pointer is a CSV
+  if (auto *GV = dyn_cast<GlobalVariable>(Ptr)) {
+    auto It = CSVAliasInfoMap.find(GV);
+    if (It != CSVAliasInfoMap.end()) {
+      // Set alias.scope and noalias metadata
+      I.setMetadata(LLVMContext::MD_alias_scope,
+                    MDNode::concatenate(I.getMetadata(LLVMContext::MD_noalias),
+                                        It->second.AliasSet));
+      I.setMetadata(LLVMContext::MD_noalias,
+                    MDNode::concatenate(I.getMetadata(LLVMContext::MD_noalias),
+                                        It->second.NoAliasSet));
+      return;
+    }
+  }
+
+  // It's not a CSV memory access, set noalias info
+  MDNode *MemoryAliasSet = MDNode::get(I.getContext(), AllCSVScopes);
+  I.setMetadata(LLVMContext::MD_noalias,
+                MDNode::concatenate(I.getMetadata(LLVMContext::MD_noalias),
+                                    MemoryAliasSet));
+}
+
+template<bool SegregateStackAccesses>
+PreservedAnalyses
+CSVAAP<SegregateStackAccesses>::run(Function &F, FunctionAnalysisManager &FAM) {
+  // Get the result of the GCBI analysis
+  auto &M = *F.getParent();
+  auto &MAM = FAM.getResult<ModuleAnalysisManagerFunctionProxy>(F).getManager();
+  GCBI = MAM.getCachedResult<GeneratedCodeBasicInfoAnalysis>(M);
+  if (!GCBI)
+    GCBI = &(FAM.getResult<GeneratedCodeBasicInfoAnalysis>(F));
+  revng_assert(GCBI != nullptr);
+
+  // Initialize the alias information for the CSVs.
+  initializeAliasInfo(M);
+
+  // Create the buckets in order to differentiate the accesses onto the stack
+  // from the non ones. Meanwhile, pointer arithmetic done with add instruction
+  // is transformed to use a getelementptr. Then, decorate the load/store
+  // operations with the new alias information.
+  if constexpr (SegregateStackAccesses) {
+    SegregateDirectStackAccesses SDSA(
+      static_cast<CSVAliasAnalysisInterface *>(this));
+    SDSA.segregateAccesses(F);
+  }
+
+  // Further decorate the IR with the alias information for the CSVs. Preserve
+  // the information on the stack accesses if previously added.
+  for (BasicBlock &BB : F)
+    for (Instruction &I : BB)
+      decorateMemoryAccesses(I);
+
+  return PreservedAnalyses::none();
+}
+
+using SDSA = SegregateDirectStackAccesses;
+
+void SDSA::segregateAccesses(Function &F) {
+  Value *SP = nullptr;
```

Comment:
Why not the following?

```suggestion
  Value *SP = CSVAA->GCBI->spReg();
```

Reply:
Yep, probably a leftover.

Reply:
Not the same thing. `SP` points to the instruction whose pointer operand is `@rsp`, not to the CSV `GlobalVariable *` (which is not even assigned to a variable here).

Reply:
Consider rewriting this piece of code from scratch taking inspiration from the old one.
I think we had incremental fixes here that make it hard to reason about it.

Reply:
Yeah, this needs to be re-written.